### DISCUSSION TITLE
Sensing + iOS compability

### DISCUSSION THREAD

**Discussion by Eaielectronic** [TurboWarp/extensions#1690] [Ideas]
Would it be possible to create an extension with the JavaScript Touch events API to create a version of sensing + compatible with iOS and any mobile device. I have already tried a code that returns positions and number of touches.
`// Définition de l'extension multi-touch
class MultiTouchExtension {
    constructor() {
        // Initialise une liste des touches actives
        this.activeTouches = [];
        
        // Ajoute les écouteurs d'événements tactiles
        window.addEventListener('touchstart', this.handleTouchStart.bind(this));
        window.addEventListener('touchmove', this.handleTouchMove.bind(this));
        window.addEventListener('touchend', this.handleTouchEnd.bind(this));
    }

    // Fonction appelée lorsqu'un doigt touche l'écran
    handleTouchStart(event) {
        this.updateTouches(event.touches);
    }

    // Fonction appelée lorsqu'un doigt se déplace sur l'écran
    handleTouchMove(event) {
        this.updateTouches(event.touches);
    }

    // Fonction appelée lorsqu'un doigt est retiré de l'écran
    handleTouchEnd(event) {
        this.updateTouches(event.touches);
    }

    // Mise à jour des touches actives
    updateTouches(touches) {
        this.activeTouches = [];
        for (let i = 0; i < touches.length; i++) {
            this.activeTouches.push({
                x: touches[i].clientX,
                y: touches[i].clientY
            });
        }
    }

    // Bloc pour obtenir le nombre de touches actives
    getTouchCount() {
        return this.activeTouches.length;
    }

    // Bloc pour obtenir les coordonnées X d'une touche spécifique
    getTouchX(args) {
        const index = args.INDEX - 1;
        if (index >= 0 && index < this.activeTouches.length) {
            return this.activeTouches[index].x;
        }
        return 0;
    }

    // Bloc pour obtenir les coordonnées Y d'une touche spécifique
    getTouchY(args) {
        const index = args.INDEX - 1;
        if (index >= 0 && index < this.activeTouches.length) {
            return this.activeTouches[index].y;
        }
        return 0;
    }

    // Définition des blocs Scratch
    getInfo() {
        return {
            id: 'multitouch',
            name: 'Multi-Touch',
            blocks: [
                {
                    opcode: 'getTouchCount',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'nombre de touches',
                },
                {
                    opcode: 'getTouchX',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'x de la touche [INDEX]',
                    arguments: {
                        INDEX: {
                            type: Scratch.ArgumentType.NUMBER,
                            defaultValue: 1
                        }
                    }
                },
                {
                    opcode: 'getTouchY',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'y de la touche [INDEX]',
                    arguments: {
                        INDEX: {
                            type: Scratch.ArgumentType.NUMBER,
                            defaultValue: 1
                        }
                    }
                }
            ]
        };
    }
}

// Enregistrement de l'extension dans TurboWarp
Scratch.extensions.register(new MultiTouchExtension()); `

**Comment by Eaielectronic** [comment]
`
class MultiTouchExtension {
    constructor() {
        // Initialise une liste des touches actives
        this.activeTouches = [];
        
        window.addEventListener('touchstart', this.handleTouchStart.bind(this));
        window.addEventListener('touchmove', this.handleTouchMove.bind(this));
        window.addEventListener('touchend', this.handleTouchEnd.bind(this));
    }

    // Fonction appelée lorsqu'un doigt touche l'écran
    handleTouchStart(event) {
        this.updateTouches(event.touches);
    }

    // Fonction appelée lorsqu'un doigt se déplace sur l'écran
    handleTouchMove(event) {
        this.updateTouches(event.touches);
    }

    // Fonction appelée lorsqu'un doigt est retiré de l'écran
    handleTouchEnd(event) {
        this.updateTouches(event.touches);
    }

    // Mise à jour des touches actives
    updateTouches(touches) {
        this.activeTouches = [];
        for (let i = 0; i < touches.length; i++) {
            this.activeTouches.push({
                x: touches[i].clientX,
                y: touches[i].clientY
            });
        }
    }

    // Bloc pour obtenir le nombre de touches actives
    getTouchCount() {
        return this.activeTouches.length;
    }

    // Bloc pour obtenir les coordonnées X d'une touche spécifique
    getTouchX(args) {
        const index = args.INDEX - 1;
        if (index >= 0 && index < this.activeTouches.length) {
            return this.activeTouches[index].x;
        }
        return 0;
    }

    // Bloc pour obtenir les coordonnées Y d'une touche spécifique
    getTouchY(args) {
        const index = args.INDEX - 1;
        if (index >= 0 && index < this.activeTouches.length) {
            return this.activeTouches[index].y;
        }
        return 0;
    }

    // Définition des blocs Scratch
    getInfo() {
        return {
            id: 'multitouch',
            name: 'Multi-Touch',
            blocks: [
                {
                    opcode: 'getTouchCount',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'nombre de touches',
                },
                {
                    opcode: 'getTouchX',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'x de la touche [INDEX]',
                    arguments: {
                        INDEX: {
                            type: Scratch.ArgumentType.NUMBER,
                            defaultValue: 1
                        }
                    }
                },
                {
                    opcode: 'getTouchY',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'y de la touche [INDEX]',
                    arguments: {
                        INDEX: {
                            type: Scratch.ArgumentType.NUMBER,
                            defaultValue: 1
                        }
                    }
                }
            ]
        };
    }
}
Scratch.extensions.register(new MultiTouchExtension()); `

**Comment by Skyhigh173** [comment]
there will be an extension called Mobile Touch.
